tfx/components/example_gen/big_query_example_gen/executor_test.py

```diff
@@ -64,9 +67,13 @@ def setUp(self):
     # Mock BigQuery result schema.
     self._schema = [
         bigquery.SchemaField('i', 'INTEGER', mode='REQUIRED'),
+        bigquery.SchemaField('i2', 'INTEGER', mode='REQUIRED'),
```

Comment:
quick question, when BigQuery returns a list, is the scheme still "INTEGER"

Reply:
I think so. I don't exactly know where bigquery.SchemaField came from, but if you study the results.schema you find `SchemaField('array_agg', 'INTEGER', 'REPEATED', None, ())`. It could be that `mode='REPEATED'` though.

```
query = """
SELECT ARRAY_AGG(x) AS array_agg
FROM UNNEST([2, 1, -2, 3, -2, 1, 2]) AS x;
"""


client = bigquery.Client()
# Dummy query to get the type information for each field.
query_job = client.query(query)
results = query_job.result()
type_map = {}

for field in results.schema:
    print(field)
```

Reply:
It can apparently also be NULLABLE `SchemaField('array_agg', 'STRING', 'NULLABLE', None, ())`
```
query = """
SELECT FORMAT("%T", ARRAY_AGG(x)) AS array_agg
FROM UNNEST([2, 1, -2, 3, -2, 1, 2]) AS x;
"""


client = bigquery.Client()
# Dummy query to get the type information for each field.
query_job = client.query(query)
results = query_job.result()
type_map = {}

for field in results.schema:
    print(field)
```

Reply:
> It can apparently also be NULLABLE `SchemaField('array_agg', 'STRING', 'NULLABLE', None, ())`

handling None is tricky, as different users might want different behaviors (empty feature or feature not exist), if data contains None, using ImportExampleGen can provide the desired behavior as the tf.Examples is created by users

Reply:
> I think so. I don't exactly know where bigquery.SchemaField came from, but if you study the results.schema you find `SchemaField('array_agg', 'INTEGER', 'REPEATED', None, ())`. [In the current implementation](https://github.com/tensorflow/tfx/blob/b66ad84e9fd216b78bc72ea5bc731d027a3b3f4d/tfx/components/example_gen/big_query_example_gen/executor.py#L42) this is how the type is derived.
> 
> It could be that `mode='REPEATED'` though. It is not intuitively obvious that a repeated/array field can't be nullable or required. [This one](https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#TableFieldSchema.FIELDS.mode) however says it one of them.

Is that possible to check this REPEATED instead of isinstance(xx, list)?

Reply:
You mean NULLABLE and not None? The None is likely the description and the empty tuple is likely fields ([docs](https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#TableFieldSchema.FIELDS.mode)).

In this case NULLABLE seems to refer to that the entire array could be missing (not that the array is empty.
```
SELECT ARRAY_AGG(x) AS array_agg
FROM UNNEST([]) AS x;
```
`SchemaField('array_agg', 'INTEGER', 'REPEATED', None, ())`

The schema is lost at that point in the current structure of the program (`__init__` vs `RowToExample`). We could however assign it the type to "REPEATED_INTEGER" or something. The question is that is an improvement.

Another question is also, does BigQuery allow for array of arrays and other corner cases?

Reply:
Scheme is returned by BigQuery, we can not set it (except in test, we fake the BigQuery result)

Reply:
@1025KB : Sorry, I was vague and I rephrased it now. I meant that we could create a new type in our `_type_map`

I tried the solution of handling all the arrays in one place, but I don't think it is better than what we have now.

```
    for key, values in instance.items():
      data_type = self._type_map[key]
      if values is None:
        feature[key] = tf.train.Feature()
      else:
        # Handle arrays and single values in the same way
        if not isinstance(values, list):
          values = [values]
        elif data_type in ('INTEGER', 'BOOLEAN'):
          feature[key] = tf.train.Feature(
              int64_list=tf.train.Int64List(value=values))
        elif data_type == 'FLOAT':
          feature[key] = tf.train.Feature(
              float_list=tf.train.FloatList(value=values))
        elif data_type == 'STRING':
          feature[key] = tf.train.Feature(
              bytes_list=tf.train.BytesList(value=[tf.compat.as_bytes(value) for value in values]))
        else:
          # TODO(jyzhao): support more types.
          raise RuntimeError(
              'BigQuery column type {} is not supported.'.format(data_type))
    return tf.train.Example(features=tf.train.Features(feature=feature))
```

Reply:
I also updated the test so that the mode is more realistic

Could you elaborate on ImportExampleGen and what you think would be the suitable thing to do if we get an empty array. I guess that if the array is NULL, it would be handled the same way as when an integer is null.